shell / exec: Improve error messages to users on exec failures

chu11 opened the pull request:
Quick attempt to improve error messages per discussion in #2481.

BEFORE

```
>flux mini run nosuchprocess
0.451s: flux-shell[0]: FATAL: failed to start taskid=0
0.453s: job.exception type=exec severity=0 failed to start taskid=0
flux-job: task(s) exited with exit code 1

>flux exec nosuchprocess
flux-exec: Error: rank 0: No such file or directory: Success
```

AFTER

```
>flux mini run nosuchprocess
0.249s: flux-shell[0]: FATAL: task 0 failed: nosuchprocess: No such file or directory
0.251s: job.exception type=exec severity=0 task 0 failed: nosuchprocess: No such file or directory
flux-job: task(s) exited with exit code 1

>flux exec nosuchprocess
flux-exec: Error: rank 0: nosuchprocess: No such file or directory
```

I debated how to tweak the one in the shell.  I didn't want to output `exec failed` as suggested by @grondo, b/c that might suggest `execvp()` failed, but technically there could be another failure cause (hypothetically `dup2()` fails due to file descriptor max being hit).  But maybe that's just me, perhaps the word `exec` is generic enough to not necessarily suggest `execvp()`?

The error message in `exec` is much better, and also removes that extraneous strerror output (due to call to `log_err()` instead of `log_msg()`)

I went ahead and peeled off #2384 from #2460 into this PR as well.

chu11 commented:
Oh I didn't add any tests, but now that I think about it, it might be good to make sure that the error message of "No such file ..." is actually seen.  I'll tweak on Monday.

grondo commented:
> that might suggest execvp() failed, but technically there could be another failure cause (hypothetically dup2() fails due to file descriptor max being hit).

This argues for the original proposal, printing explicit errors to stderr in the child.

However, Looking at the error message above, I agree it is fine without the `exec failed`.

grondo commented:
Oh, actually (and I admit this is minor), it would be nice if the error message somehow indicated that the task failed before it was executed. The generic "task failed" could be misinterpreted that the task failed in some other manner. For that keeping something about "task 0: start failed" or "task 0: unable to execute" would be useful (if you don't like "exec failed")

garlick commented:
Small typo in the commit msesage of the "improve error message to the user" commit (currently 06434ef)

> an extranous errno

chu11 commented:
rebased & re-pushed fixing issues discussed above

- added / tweaked some tests
- fixed the comment typo @garlick found
- adjusted the error message per @grondo comments

chu11 commented:
FYI, the poor diff coverage is entirely due to the `flux_log()` -> `fprintf()` change in libsubprocess.  All are "impossible" error paths.

grondo commented:
Hm,  I'm not seeing the pre-exec errors on stderr for jobs. Something to look into likely unrelated to this PR. Will investigate more later, but
```console
$ flux mini run --setattr="system.cwd=/foo" pwd
/tmp
```

Should see an error about "unable to chrdir to /foo, going to /tmp instead".

chu11 commented:
> Should see an error about "unable to chrdir to /foo, going to /tmp instead".

I don't see a call to `flux_cmd_setcwd()` in the shell.  Which would definitely be the problem.  I'll write up an issue when I get in and I can attach a fix in this issue.

chu11 commented:
Repushed with an additional fix for #2679.

Perhaps the most controversial part of this fix is I changed the error message from "Couldn't change ..." to "Could not change ...".  Partially b/c "Could not" is more formal.  But mostly b/c I got sick of trying to figure out a way to get sharness test to work with a single quote, and to make sure it wasn't a bash-ism.

grondo commented:
See my comment in #2679. I don't think separately changing working directory in each task the right fix. The job-shell should also be in the working directory of the job, since the shell runs as the user. Then, until we allow separate working dirs for each task, the tasks should just inheret the shell's working dir.

chu11 commented:
> See my comment in #2679. I don't think separately changing working directory in each task the right fix. The job-shell should also be in the working directory of the job, since the shell runs as the user. Then, until we allow separate working dirs for each task, the tasks should just inheret the shell's working dir.

Yup, just saw.  Backing out my fix in this PR.  We'll leave it as just fixes for #2384 and #2481.

grondo commented:
> We'll leave it as just fixes for #2384 and #2481.

Ok, that works. Sorry about that.
